metadata/src/main/java/org/apache/kafka/timeline/SnapshotRegistry.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.timeline;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+
+/**
+ * A registry containing snapshots of timeline data structures.
+ * We generally expect a small number of snapshots-- perhaps 1 or 2 at a time.
+ * Therefore, we use ArrayLists here rather than a data structure with higher overhead.
+ */
+public class SnapshotRegistry {
+    class SnapshotIterator implements ListIterator<Snapshot> {
+        private Snapshot cur;
+        private Snapshot lastResult = null;
+
+        SnapshotIterator(Snapshot startAfter) {
+            this.cur = startAfter;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return cur.next() != head;
+        }
+
+        @Override
+        public Snapshot next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            cur = cur.next();
+            lastResult = cur;
+            return cur;
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            return cur != head;
+        }
+
+        @Override
+        public Snapshot previous() {
+            if (!hasPrevious()) {
+                throw new NoSuchElementException();
+            }
+            Snapshot result = cur;
```

Comment:
This is a bit weird with the input param startAfter, which suggests iterating all items after startAfter. With previous, we are returning items before and including startAfter.

Reply:
ListIterator supports going backwards.  startAfter just indicates that the iteration starts after the given point

Reply:
In that case, we should not return the snapshot at startAfter, right?

Reply:
Hmm, I'm not sure I follow the question.  I might be missing something.

startAfter just means start after X.  So if your list is A, B, C, D and you pass startAfter = B the first call to next() will return C.

If, instead, you choose to iterate backwards using previous() then of course you will get B and then A.  But you still "started after" B, you just chose to go backwards.